### PR TITLE
Optimize numpy operations in MKDADensity Estimator and (M)KDAKernel

### PR DESCRIPTION
**Pull request by adelavega**
- Speeds in KDA meta-analysis, especially in the context of monte-carlo correction
- Main gains came from optimizing numpy operations, especially switchingout `np.unique` -> `unique_rows`
- Also chunked kernel operations (i.e. convolution & masking), by experiment ID, which is more efficient

Profile summary: Around a ~2.5-3x speed up for `correct_fwe_montecarlo` on a dataset with ~1000 coordinates

### CODE REVIEW

**Comment by adelavega** [comment]
From this issue: https://github.com/numpy/numpy/issues/11136
I found an implementation that uses pandas and hashing to get unique rows wihout sorting for 2D arrays.

It seems to be significantly faster. All in all, for a dataset w/ ~100 studies, it went from ~0.53 to ~0.13s.

Still need to test more, but this should be helpful at least.

**Comment by adelavega** [comment]
Replaced with scikit image version: https://github.com/scikit-image/scikit-image/blob/64103e6c90917fcfdef8343fd7dd4df32c910446/skimage/util/unique.py#L47

**Comment by adelavega** [comment]
Profiling results using the "neurosynth_laird_studies.json" dataset (1117 coordinaes in 17 studies).

`correct_fwe_montecarlo, with 10 iterations`:

`main` branch: 9.67 s / ~1.1 iterations/s
This PR: 4.19s / ~3 iterations/s

It's looking like a 2.5x-3x speed up, just from speeding up `compute_kda_ma`

**Comment by adelavega** [comment]
Ugh, `black` made changes in tons of files. Shouldn't black already have been run on `main`?

File: nimare/meta/cbma/mkda.py
Comment (adelavega) [inline nimare/meta/cbma/mkda.py:169]: This seems like an edge case that is not worth the slow down, so I went w/ the faster version

File: nimare/meta/utils.py
Comment (adelavega) [inline nimare/meta/utils.py:278]: This is straight from skimage, so I could import but  I don't want that as a dependency

**Comment by tsalo** [comment]
I think you must have used a different version of black, because all files _should_ have already had black run on them and the linter is currently failing.

File: nimare/meta/utils.py
Comment (tsalo) [reply nimare/meta/utils.py:278]: Is their license amenable to using their code directly? If so, then I'm all for just using their function without adding a dependency.

Comment (adelavega) [reply nimare/meta/utils.py:278]:
```
Copyright (C) 2019, the scikit-image team
All rights reserved.

Redistribution and use in source and binary forms, with or without
modification, are permitted provided that the following conditions are
met:

 1. Redistributions of source code must retain the above copyright
    notice, this list of conditions and the following disclaimer.
 2. Redistributions in binary form must reproduce the above copyright
    notice, this list of conditions and the following disclaimer in
    the documentation and/or other materials provided with the
    distribution.
 3. Neither the name of skimage nor the names of its contributors may be
    used to endorse or promote products derived from this software without
    specific prior written permission.
```

Looks like it just need to add the copyright

**Comment by adelavega** [comment]
By the way, I think the other major speed up we could potentially get is related to `sparse` memory usage.

Currently, `compute_kda_ma` returns a sparse matrix, which `correct_fwe_montecarlo` then converts to image because it calls:

```
self.kernel_transformer.transform(
            iter_df, masker=self.masker, return_type="array"
        )
```

with `return_type` == 'array`.

This means we spend a lot of time going from sparse --> dense. We save a tiny bit of memory doing this because we loop over experiments and only keep the masked voxels, but its slow. 

Starting from `dense` and staying that would should speed things up, but it's not clear to me by how much. It also seems like memory is a bigger problem than speed, so it may not be worth it.

@JulioAPeraza

File: nimare/meta/kernel.py
Comment (adelavega) [reply nimare/meta/kernel.py:208]: Apparently this check is necessary, because without it tests fails. However, it _shouldn't_ be necessary because the output should be sparse here. 

It's not in `ALEKernel` however.

This is inconsistent w/ the API because if you say "return_type='sparse'" you could get back a dense matrix w/ an ALEKernel

File: nimare/meta/utils.py
Comment (tsalo) [reply nimare/meta/utils.py:278]: Awesome! I'm happy to just use the code then.

File: nimare/meta/cbma/mkda.py
Comment (tsalo) [reply nimare/meta/cbma/mkda.py:169]: I can test it out tomorrow.

Comment (tsalo) [reply nimare/meta/cbma/mkda.py:169]: `tests.test_meta_mkda.test_MKDADensity_approximate_null` went from 7.37s to 6.49s after making the change on my MacBook Pro (OSX 12.3.1). Maybe the architecture Tal was referring to is the M1 chip? Another possibility is that the problem was unique to multiprocessing, and when I switched to joblib it resolved the issue without anyone noticing.

In case anyone wants to dig further into it, the commit where the workaround was added is 68c515a5511f9baaa265bc24026a12defe86d4fb.

**Comment by adelavega** [comment]
@tsalo I believe this is ready to go as soon as you approve

File: nimare/meta/kernel.py
Comment (tsalo) [reply nimare/meta/kernel.py]: Moved to #692.

File: nimare/meta/kernel.py
Comment (tsalo) [inline nimare/meta/kernel.py:137]:
```suggestion
            )  # TODO: Possibly optimize this
```

File: nimare/meta/kernel.py
Comment (tsalo) [inline nimare/meta/kernel.py:188]:
```suggestion
            )  # TODO: Possibly allow this to be passed in to not recompute
```

File: nimare/meta/kernel.py
Comment (tsalo) [inline nimare/meta/kernel.py:209]:
```suggestion
        # TODO: Optimize this, possibly make own function
```

File: nimare/meta/kernel.py
Comment (tsalo) [inline nimare/meta/kernel.py:420]:
```suggestion
    model_dir : :obj:`str`, optional
```

File: nimare/meta/utils.py
Comment (tsalo) [inline nimare/meta/utils.py:392]:
```suggestion
        """Convolve peaks with a spherical kernel.

        Parameters
        -----------
        kernel : 2D numpy.ndarray
            IJK coordinates of a sphere, relative to a central point (not the brain template).
        peaks : 2D numpy.ndarray
            The IJK coordinates of peaks to convolve with the kernel.

        Returns
        --------
        sphere_coords : 2D numpy.ndarray
            All coordinates that fall within any sphere.
            Coordinates from overlapping spheres will appear twice.
        """
```

Hopefully this is correct.

File: nimare/meta/utils.py
Comment (tsalo) [inline nimare/meta/utils.py:302]: Can you add a reference to the scikit-image function? E.g., ```This is taken from skimage. See :func:`skimage.util.unique_rows`.``` We'll need to add skimage to the inter sphinx mapping in `docs/conf.py`.

**Review by tsalo — CHANGES_REQUESTED** [review]
I _think_ I understand the changes to the code, and they look good to me.

I have some notes on the comments though- mostly just adding TODO to comments we might want to tackle in a future PR.

File: nimare/meta/kernel.py
Comment (adelavega) [reply nimare/meta/kernel.py:137]: Oh, sure. I forgot to remove these so I think it's probably better to just remove them.

Comment (tsalo) [reply nimare/meta/kernel.py:137]: That sounds good. I assumed they were just things you wanted to deal with later.

File: nimare/meta/utils.py
Comment (adelavega) [reply nimare/meta/utils.py:302]: Does that work w/o adding skimage as a dependency?

Comment (adelavega) [reply nimare/meta/utils.py:302]: yep, I figured it out I think.

**Comment by tsalo** [comment]
I totally forgot- can you also add `unique_rows` to `docs/api.rst`? Sorry about that!

**Comment by adelavega** [comment]
Yes, and actually I moved it to `nimare.utils` which makes more sense. We could presumably reuse it throughout the package, not just in cbma. Agree?

**Comment by tsalo** [comment]
Agreed! I'm sure there'll be a few places where it could come in handy.

**Review by tsalo — APPROVED** [review]
The changes look good to me. I might write a test or two for the new function (or just copy them from skimage) in a future PR, but I don't want to block this one any longer. Once CI passes, feel free to merge.

I'm just going to change the title slightly, to match the pattern in other PRs.

EDIT: I also realized the changes were made to the kernel and the _MKDA_ (rather than KDA) Estimator.

File: nimare/meta/kernel.py
Comment (JulioAPeraza) [reply nimare/meta/kernel.py]: Sorry that I missed this comment. I created a new folder for GitHub email and I didn't get notified of this one. I added a comment to the new issue.

**Pull request merged**
